crates/wit-component/src/linking.rs

```diff
@@ -0,0 +1,1383 @@
+    wasmparser::WASM_SYM_BINDING_WEAK,
+};
+
+mod metadata;
+
+const PAGE_SIZE_BYTES: u32 = 65536;
+// This matches the default stack size LLVM produces:
+const STACK_SIZE_BYTES: u32 = 16 * PAGE_SIZE_BYTES;
+const HEAP_ALIGNMENT_BYTES: u32 = 16;
+
+enum Address<'a> {
+    Function(u32),
+    Global(&'a str),
+}
+
+/// Represents a `dlopen`/`dlsym` lookup table enabling runtime symbol resolution
+///
+/// The top level of this table is a sorted list of library names and offsets, each pointing to a sorted list of
+/// symbol names and offsets.  See
+/// https://github.com/dicej/wasi-libc/blob/76c7e1e1cfdad577ecd7f61c67ead7a38d62a7c4/libc-top-half/musl/src/misc/dl.c
+/// for how this is used.
+struct DlOpenables<'a> {
+    /// Offset into the main module's table where function references will be stored
+    table_base: u32,
+
+    /// Offset into the main module's memory where the lookup table will be stored
+    memory_base: u32,
+
+    /// The lookup table itself
+    buffer: Vec<u8>,
+
+    /// Linear memory addresses where global variable addresses will live
+    ///
+    /// The init module will fill in the correct values at insantiation time.
+    global_addresses: Vec<(&'a str, &'a str, u32)>,
+
+    /// Number of function references to be stored in the main module's table
+    function_count: u32,
+
+    /// Linear memory address where the root of the lookup table will reside
+    ///
+    /// This can be different from `memory_base` depending on how the tree of libraries and symbols is laid out in
+    /// memory.
+    libraries_address: u32,
+}
+
+impl<'a> DlOpenables<'a> {
+    /// Construct a lookup table containing all "dlopen-able" libraries and their symbols using the specified table
+    /// and memory offsets.
+    fn new(table_base: u32, memory_base: u32, metadata: &'a [Metadata<'a>]) -> Self {
+        let mut function_count = 0;
+        let mut buffer = Vec::new();
+        let mut global_addresses = Vec::new();
+        let mut libraries = metadata
+            .iter()
+            .filter_map(|metadata| {
+                if metadata.dl_openable {
+                    let name_address = memory_base + u32::try_from(buffer.len()).unwrap();
+                    write_bytes_padded(&mut buffer, metadata.name.as_bytes());
+
+                    let mut symbols = metadata
+                        .exports
+                        .iter()
+                        .map(|export| {
+                            let name_address = memory_base + u32::try_from(buffer.len()).unwrap();
+                            write_bytes_padded(&mut buffer, export.key.name.as_bytes());
+
+                            let address = if let Type::Function(_) = &export.key.ty {
+                                Address::Function(
+                                    table_base + get_and_increment(&mut function_count),
+                                )
+                            } else {
+                                Address::Global(export.key.name)
```

Comment:
`linking::metadata::Type` only has two variants: `Function` and `Global`, so there's nothing to assert for now.  I'll change this to a `match`, though, which will give us an exhaustiveness check in case we add more variants later.